File: server/datasets/python_source-code/Multiple-G2/structural_27_avl_tree.py
class Node_2:
    def __init__(self, v_2):
        self.v_2 = v_2
        self.r_2 = None
        self.l_2 = None
        self.height = 1

def h_2(node):
    if not node:
        return 0
    return node.height

def b_2(node):
    if not node:
        return 0
    return h_2(node.l_2) - h_2(node.r_2)

def rr_2(y):
    x = y.l_2
    sub = x.r_2
    x.r_2 = y
    y.l_2 = sub
    y.height = 1 + max(h_2(y.l_2), h_2(y.r_2))
    x.height = 1 + max(h_2(x.l_2), h_2(x.r_2))
    return x

def rl_2(x):
    y = x.r_2
    sub = y.l_2
    y.l_2 = x
    x.r_2 = sub
    x.height = 1 + max(h_2(x.l_2), h_2(x.r_2))
    y.height = 1 + max(h_2(y.l_2), h_2(y.r_2))
    return y

def ins_2(rt_2, v_2):
    if not rt_2:
        return Node_2(v_2)
    if v_2 < rt_2.v_2:
        rt_2.l_2 = ins_2(rt_2.l_2, v_2)
    elif v_2 > rt_2.v_2:
        rt_2.r_2 = ins_2(rt_2.r_2, v_2)
    else:
        return rt_2

    rt_2.height = 1 + max(h_2(rt_2.l_2), h_2(rt_2.r_2))
    bf_2 = b_2(rt_2)

    if bf_2 > 1 and v_2 < rt_2.l_2.v_2:
        return rr_2(rt_2)
    if bf_2 < -1 and v_2 > rt_2.r_2.v_2:
        return rl_2(rt_2)
    if bf_2 > 1 and v_2 > rt_2.l_2.v_2:
        rt_2.l_2 = rl_2(rt_2.l_2)
        return rr_2(rt_2)
    if bf_2 < -1 and v_2 < rt_2.r_2.v_2:
        rt_2.r_2 = rr_2(rt_2.r_2)
        return rl_2(rt_2)

    return rt_2

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_g2/structural_27_avl_tree.java
class Node_2 {
    int k_2, h_2;
    Node_2 left, right;
    Node_2(int d) {
        k_2 = d;
        h_2 = 1;
    }
}
public class Tree_2 {
    public static void main(String[] args) {
        Tree_2 t = new Tree_2();
        Node_2 r = null;
        r = t.ins_2(r, 10);
    }
    int max(int a, int b) {
        return (a > b) ? a : b;
    }
    int h_2(Node_2 N) {
        if (N == null) return 0;
        return N.h_2;
    }
    int gb_2(Node_2 N) {
        if (N == null) return 0;
        return h_2(N.left) - h_2(N.right);
    }
    Node_2 upd_2(Node_2 n_2) {
        n_2.h_2 = 1 + max(h_2(n_2.left), h_2(n_2.right));
        return n_2;
    }
    Node_2 rr_2(Node_2 y) {
        Node_2 x = y.left;
        Node_2 T2 = x.right;
        x.right = y;
        y.left = T2;
        upd_2(y);
        upd_2(x);
        return x;
    }
    Node_2 rl_2(Node_2 x) {
        Node_2 y = x.right;
        Node_2 T2 = y.left;
        y.left = x;
        x.right = T2;
        upd_2(x);
        upd_2(y);
        return y;
    }
    Node_2 ins_2(Node_2 n_2, int k_2) {
        if (n_2 == null) return (new Node_2(k_2));
        if (k_2 < n_2.k_2) n_2.left = ins_2(n_2.left, k_2);
        else if (k_2 > n_2.k_2) n_2.right = ins_2(n_2.right, k_2);
        else return n_2;
        upd_2(n_2);
        int b_2 = gb_2(n_2);
        if (b_2 > 1 && k_2 < n_2.left.k_2) return rr_2(n_2);
        if (b_2 < -1 && k_2 > n_2.right.k_2) return rl_2(n_2);
        if (b_2 > 1 && k_2 > n_2.left.k_2) {
            n_2.left = rl_2(n_2.left);
            return rr_2(n_2);
        }
        if (b_2 < -1 && k_2 < n_2.right.k_2) {
            n_2.right = rr_2(n_2.right);
            return rl_2(n_2);
        }
        return n_2;
    }
}